packages/react/src/components/ComboBox/ComboBox.js

```diff
@@ -65,490 +70,479 @@ const findHighlightedIndex = ({ items, itemToString }, inputValue) => {
-    if (!doneInitialSelectedItem || prevSelectedItem !== selectedItem) {
-      return {
-        doneInitialSelectedItem: true,
-        prevSelectedItem: selectedItem,
-        inputValue: getInputValue(nextProps, state),
-      };
-    }
-    return null;
-  }
-
-  constructor(props) {
-    super(props);
-
-    this.comboBoxInstanceId = getInstanceId();
-    this.state = {
-      inputValue: getInputValue(props, {}),
-    };
-    this.textInput = React.createRef();
+const ComboBox = (props) => {
+  const {
+    ariaLabel,
+    className: containerClassName,
+    direction,
+    disabled,
+    downshiftProps,
+    helperText,
+    id,
+    initialSelectedItem,
+    invalid,
+    invalidText,
+    items,
+    itemToElement,
+    itemToString,
+    light,
+    onChange,
+    onInputChange,
+    onToggleClick, // eslint-disable-line no-unused-vars
+    placeholder,
+    selectedItem,
+    shouldFilterItem,
+    size,
+    titleText,
+    translateWithId,
+    type, // eslint-disable-line no-unused-vars
+    warn,
+    warnText,
+    ...rest
+  } = props;
+
+  const textInput = useRef();
+  const comboBoxInstanceId = getInstanceId();
+  const [inputValue, setInputValue] = useState(
+    getInputValue({
+      initialSelectedItem,
+      inputValue: '',
+      itemToString,
+      selectedItem,
+    })
+  );
+  const [prevSelectedItem, setPrevSelectedItem] = useState(null);
+  const [doneInitialSelectedItem, setDoneInitialSelectedItem] = useState(null);
+
+  if (!doneInitialSelectedItem || prevSelectedItem !== selectedItem) {
+    setDoneInitialSelectedItem(true);
+    setPrevSelectedItem(selectedItem);
+    setInputValue(
+      getInputValue({
+        initialSelectedItem,
+        inputValue,
+        itemToString,
+        selectedItem,
+      })
+    );
   }
 
-  filterItems = (items, itemToString, inputValue) =>
+  const filterItems = (items, itemToString, inputValue) =>
     items.filter((item) =>
-      this.props.shouldFilterItem({
+      shouldFilterItem({
         item,
         itemToString,
         inputValue,
       })
     );
 
-  handleOnChange = (selectedItem) => {
-    if (this.props.onChange) {
-      this.props.onChange({ selectedItem });
+  const handleOnChange = (selectedItem) => {
+    if (onChange) {
+      onChange({ selectedItem });
     }
   };
 
-  handleOnInputValueChange = (inputValue) => {
-    const { onInputChange } = this.props;
-
-    this.setState(
-      () => ({
-        // Default to empty string if we have a false-y `inputValue`
-        inputValue: inputValue || '',
-      }),
-      () => {
-        if (onInputChange) {
-          onInputChange(inputValue);
-        }
-      }
-    );
+  const handleOnInputValueChange = (inputValue) => {
+    setInputValue(inputValue || '');
   };
 
-  handleSelectionClear = () => {
-    if (this.textInput?.current) {
-      this.textInput.current.focus();
+  useEffect(() => {
+    if (onInputChange) {
+      onInputChange(inputValue);
+    }
+  });
```

Comment:
Without a dependency array, this will trigger the `onInputChange` with _every_ render.

This should only be triggered when the `inputValue` changes
```js
  useEffect(() => {
    if (onInputChange) {
      onInputChange(inputValue);
    }
  }, [inputValue]); // <-- it should have this!
```

Reply:
related #8306 #8308